go/logic/topology_recovery.go

```diff
@@ -2225,3 +2220,29 @@ func GracefulMasterTakeover(clusterName string, designatedKey *inst.InstanceKey,
 
 	return topologyRecovery, promotedMasterCoordinates, err
 }
+
+func gracefulMasterTakeover(demotedMasterSelfBinlogCoordinates *inst.BinlogCoordinates, designatedInstance *inst.Instance, analysisEntry inst.ReplicationAnalysis) (topologyRecovery *TopologyRecovery, promotedMasterCoordinates *inst.BinlogCoordinates, err error) {
+	log.Infof("GracefulMasterTakeover: Will wait for %+v to reach master coordinates %+v", designatedInstance.Key, *demotedMasterSelfBinlogCoordinates)
+	if designatedInstance, _, err = inst.WaitForExecBinlogCoordinatesToReach(&designatedInstance.Key, demotedMasterSelfBinlogCoordinates, time.Duration(config.Config.ReasonableMaintenanceReplicationLagSeconds)*time.Second); err != nil {
+		return nil, nil, err
+	}
+	promotedMasterCoordinates = &designatedInstance.SelfBinlogCoordinates
+
+	log.Infof("GracefulMasterTakeover: attempting recovery")
+	recoveryAttempted, topologyRecovery, err := ForceExecuteRecovery(analysisEntry, &designatedInstance.Key, false)
```

Comment:
Hi @o-fedorov , I'm not sure about this error propagation and handling. In the original flow, if `ForceExecuteRecovery` returned err, but also returned `recoveryAttempted` and `topologyRecovery`, `GracefulMasterTakeover` continued.
Now it makes master R/W (which is fine), but then returns.

Reply:
Thank you, @kamil-holubicki , I forgot to clear the error at the end of the function. Updated now.

Note that in the original code, `err` is ignored below the refactored part, and is eventually overridden.